### PR TITLE
Implement async calls

### PR DESCRIPTION
**Pull request by bbyalcinkaya**
This PR implements intra-shard async calls.

### CODE REVIEW

File: elrond-config.md [hunk 1]
```diff
@@ -414,18 +416,36 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
-         <vmOutput> _ => VMOutput( OK , .Bytes , OUT , LOGS) </vmOutput>
       [priority(60)]
 
     syntax InternalCmd ::= "#waitWasm"  [klabel(#waitWasm), symbol]
  // ----------------------------------
     rule <commands> #waitWasm => .K ... </commands>
          <instrs> .K </instrs>
       [priority(60)]
+
+    syntax InternalInstr ::= "#waitCommands"
```
Comment (virgil-serbanuta) [inline elrond-config.md:429]: Would it be hard to add somewhere a description of what these wait commands do and now they interact with each other? Or a description of how execution works, given that we have three execution cells (<k>, <commands> and <instrs>)?

File: elrond-config.md [hunk 2]
```diff
@@ -414,18 +416,36 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
+    rule <commands> #transferSuccess => .K ... </commands>
+         <instrs> #waitCommands ... </instrs>
+    
+
+    syntax InternalCmd ::= "#setVMOutput"
+ // -------------------------------------
+    rule <commands> #setVMOutput => .K ... </commands>
+         <out> OUT </out>
+         <logs> LOGS </logs>
+         <outputAccounts> OUT_ACCS</outputAccounts>
```
Comment (virgil-serbanuta) [inline elrond-config.md:445]: space after variable?

File: elrond-config.md [hunk 3]
```diff
@@ -414,18 +416,36 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
+         <vmOutput> _ => VMOutput( OK , .Bytes , OUT , LOGS , OUT_ACCS) </vmOutput>
```
Comment (virgil-serbanuta) [inline elrond-config.md:446]: Either remove the space after '(' or add one before ')'.

This applies everywhere, I guess.

File: elrond-node.md [hunk 4]
```diff
@@ -105,15 +114,53 @@ Storage maps byte arrays to byte arrays.
...
                         [klabel(VMOutput), symbol]
 
- // ------------------------------------------------------------------
+    syntax OutputAccount ::= OutputAccount ( addr: Bytes, transfers: List )             [klabel(OutputAccount), symbol]
+    syntax OutputTransfer ::= OutputTransfer ( sender: Bytes, value: TransferValue )    [klabel(OutputTransfer), symbol]
+
+    syntax Bool ::= nonZeroOutputTransfer(OutputTransfer)      [function, total]
+ // ---------------------------------------------------------------------------
+    rule nonZeroOutputTransfer(OutputTransfer(_, I:Int))                 => I =/=Int 0
+    rule nonZeroOutputTransfer(OutputTransfer(_, esdtTransfer(_, I, _))) => I =/=Int 0
```
Comment (virgil-serbanuta) [inline elrond-node.md:126]: I'm curious: would it be possible to have `I <Int 0`? Either in a test, or in a real world example.

**Review by virgil-serbanuta — COMMENTED** [review]
I didn't manage to go through the entire PR yet.

File: vmhooks/async.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    syntax InternalCmd ::= "#executeAsyncLocalCallback"        [symbol(executeAsyncLocalCallback)]
+ // ---------------------------------------------------------------------------------------------
+    rule [executeAsyncLocalCallback]:
+        <commands> #executeAsyncLocalCallback
+                => callContractCb(PARENT, chooseCallback(RC, CALL),
+                                <vmInput>
+                                  <caller> CHILD </caller>
+                                  <callArgs> argsForCallback(RC, MSG, OUT) </callArgs>
+                                  <callValue> extractLastValue(PARENT, VMOUTPUT) </callValue>
```
Comment (virgil-serbanuta) [inline vmhooks/async.md:149]: What does this do/mean?

File: vmhooks/managedei.md [hunk 6]
```diff
@@ -140,6 +140,69 @@ module MANAGEDEI
...
+            <callState>
+              <asyncCalls> ListItem( #asyncCall( ... closure: CLOSURE )) ... </asyncCalls>
+              ...
+            </callState>
+          ) ...
+        </callStack>
+
+    rule [managedGetCallbackClosure-err]:
+        <instrs> hostCall ( "env" , "managedGetCallbackClosure" , [ i32 .ValTypes ] -> [ .ValTypes ] )
+              => #throwException(ExecutionFailed, "no callback for closure, cannot call callback directly")
```
Comment (virgil-serbanuta) [inline vmhooks/managedei.md:198]: Hmmmm... So, if there is an entry in the async calls list, would it be possible to call the callback directly? Or am I misunderstanding this?

File: data/async-call.k [hunk 7]
```diff
@@ -0,0 +1,56 @@
+requires "list-bytes.k"
+
+module ASYNC-CALL
+  imports private BYTES
+  imports private LIST-BYTES
+  imports private INT-SYNTAX
+  imports private BASIC-K
+
+  syntax AsyncCall ::= #asyncCall(
+                          status: AsyncCallStatus,
```
Comment (virgil-serbanuta) [inline data/async-call.k:10]: Consider removing if not needed

File: vmhooks/async.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        </vmOutput>
+
+    syntax String ::= chooseCallback(ReturnCode, AsyncCall)    [function, total]
+ // -------------------------------------------------------------------------------
+    rule chooseCallback(OK,              #asyncCall(... successCallback: F)) => F
+    rule chooseCallback(_:ExceptionCode, #asyncCall(... errorCallback: F)) => F
+
+    syntax ListBytes ::= argsForCallback(ReturnCode, Bytes, ListBytes)    [function, total]
+ // -------------------------------------------------------------------
+    rule argsForCallback(OK, _, OUT) => ListItem(wrap(Int2Bytes(0, BE, Unsigned))) OUT
```
Comment (virgil-serbanuta) [inline vmhooks/async.md:227]: Add a comment specifying what the first arg is?

**Review by virgil-serbanuta — APPROVED** [review]
I think that there are still two unanswered questions, if you can, please add answers before merging.

File: elrond-node.md (hunk 4 above)
Comment (bbyalcinkaya) [reply elrond-node.md:143]: These values are parsed from buffers as unsigned ints, so it is not possible.

File: vmhooks/managedei.md (hunk 6 above)
Comment (bbyalcinkaya) [reply vmhooks/managedei.md]: That's a good catch, thanks! Added `<callType> AsynchronousCallBack </callType>` to fix it.

**Pull request merged**